### PR TITLE
WIP: Automatic Validator count adjust with dampening

### PR DESCRIPTION
**Pull request by g2udevelopment**
This PR build on top of the previous work of @shawntabrizi. It implements the features in #6997. The goal is to have an automatic adjustment of the validator count when a new era starts if certain conditions are matched.
It also adds a flag to turn on/off the use of the dampening function and it adds an extra MaxValidatorCount storage item.
On top of that I added mocks and a few tests, to test the basic functioning of the dampening function.

I think a few test could be added so it would be helpfull if someone can add them. Also validation of the rules is still needed.

### CODE REVIEW

**Review by kianenigma — CHANGES_REQUESTED** [review]
I disagree with the logic of validator count update being baked into staking. 

With `ValidatorCountAdjust `, you allow a user of substrate to opt out of this feature, which is good. But what if someone wants to have some update function, like the one proposed for kusama, but slightly different? this won't be possible with your API. 

As recommended, I think having a `type ValidatorConvert: Convert<u32, u32>`. For a chain that wants to opt out, they will pass in an `IdentityConvert` which doesn't touch the count. Kusama can have its own custom implementation, and an other chain is free to customize this.

File: frame/staking/src/lib.rs [hunk 1]
```diff
@@ -3127,6 +3139,42 @@ impl<T: Trait> Module<T> {
 		})
 	}
 
+	/// Use a damping function to update the target number of validators for the next selection.
+	fn update_target_validators(mut exposures: Vec<BalanceOf<T>>) -> u32 {
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:3143]: the name exposure should ideally refer to the type `Exposure`, maybe call this `exposed_balances`

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -3127,6 +3139,42 @@ impl<T: Trait> Module<T> {
...
+		let average = |all: &[BalanceOf<T>]| -> BalanceOf<T> {
+			all.iter().fold(Zero::zero(), |total: BalanceOf<T>, x: &BalanceOf<T>| {
+				// `total` should never saturate since total issuance fits inside `Balance`
+				total.saturating_add(*x)
+			}) / BalanceOf::<T>::saturated_from(all.len() as u128)
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:3148]: I think we assume `x.len() as u32` in runtime and then you can just `into()` it to `BalanceOf<T>`.

Comment (kianenigma) [reply frame/staking/src/lib.rs:3148]: i.e. 

```
```suggestion
			}) / BalanceOf::<T>::from(all.len() as u32)
```

Comment (kianenigma) [reply frame/staking/src/lib.rs:3148]: would be good to be safe against len == 0 as well.

File: frame/staking/src/tests.rs [hunk 3]
```diff
@@ -346,6 +346,78 @@ fn staking_should_work() {
...
 
+#[test]
+fn maximum_validator_set() {
+	ExtBuilder::default()
+		.maximum_validator_count(4)
+		.build()
+		.execute_with(|| {
+			assert_eq!(Staking::maximum_validator_count(), 4);
+		});
+}
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:357]:
```suggestion
#[test]
fn maximum_validator_set() {
	ExtBuilder::default()
		.maximum_validator_count(4)
		.build()
		.execute_with(|| {
			assert_eq!(Staking::maximum_validator_count(), 4);
		});
}
```

superfluous, this is not testing the logic of the code, this is testing the setup of the tests.

Comment (kianenigma) [reply frame/staking/src/tests.rs:357]: We have one test called `basic_setup_works()` or something along those lines in which you can test this.

**Comment by g2udevelopment** [comment]
> I disagree with the logic of validator count update being baked into staking.
> 
> With `ValidatorCountAdjust `, you allow a user of substrate to opt out of this feature, which is good. But what if someone wants to have some update function, like the one proposed for kusama, but slightly different? this won't be possible with your API.
> 
> As recommended, I think having a `type ValidatorConvert: Convert<u32, u32>`. For a chain that wants to opt out, they will pass in an `IdentityConvert` which doesn't touch the count. Kusama can have its own custom implementation, and an other chain is free to customize this.

Sorry I missed the conversation about baking in the function in staking in the first commits. I think this is a good suggestion.

So I will remove the boolean flag , add the `Convert` type and provide both an Identity function and the current implementation. Then the Identity Implementation will serve as the `bool` false in the current implementation.

**Comment by kianenigma** [comment]
> add the Convert type and provide both an Identity function. 

Yes, and this is already implemented in substrate 

https://github.com/paritytech/substrate/blob/4be954a8463e6f33e0ec854c4b8bcd940af260fc/primitives/runtime/src/traits.rs#L222-L226

So use this for identity.

**Comment by gnunicorn** [comment]
Closed because it's been stale for a while.

**Pull request closed without merging**